File: src/event_register.py
"""Cursor Layer of the OpusManager and the classes required to make it work"""
from __future__ import annotations
from typing import List, Optional, Tuple, Union, Any
from collections.abc import Callable
from .opusmanager.miditree import MIDITree, MIDITreeEvent

class EventRegister:
    RADIX = 12
    def __init__(self):
        self.register: Optional[MIDITreeEvent] = None

    def add_digit(self, value: int) -> None:
        """Pass a digit to the register"""

        if self.register is None:
            self.register = MIDITreeEvent(
                value,
                base=self.RADIX,
                relative=False
            )
        elif self.register.relative:
            self.register.note *= value
        else:
            self.register.note *= self.RADIX
            self.register.note += value

    def is_ready(self) -> bool:
        # If the register is ready, apply it
        return self.register is not None \
            and (
                self.register.relative \
                or self.register.note >= self.RADIX
            )

    def clear(self) -> None:
        """Cancel event being input."""
        self.register = None

    def fetch(self) -> Optional[MIDITreeEvent]:
        """Get the register, unsetting it"""
        output = self.register
        self.register = None
        return output

    def relative_add_entry(self) -> None:
        """Let the register know the user is inputting a relative event, moving up some keys"""
        self.register = MIDITreeEvent(1, base=self.RADIX, relative=True)

    def relative_subtract_entry(self) -> None:
        """Let the register know the user is inputting a relative event, moving down some keys"""
        self.register = MIDITreeEvent(-1, base=self.RADIX, relative=True)

    def relative_downshift_entry(self) -> None:
        """Let the register know the user is inputting a relative event, shifting down octaves"""
        self.register = MIDITreeEvent(-1 * self.RADIX, base=self.RADIX, relative=True)

    def relative_upshift_entry(self) -> None:
        """Let the register know the user is inputting a relative event, shifting up octaves"""
        self.register = MIDITreeEvent(self.RADIX, base=self.RADIX, relative=True)

File: kotlin/app/src/main/kotlin/radixulous/app/event_register.kt
package radixulous.app
import radixulous.app.opusmanager.OpusEvent

class EventRegister {
    var RADIX = 12
    var register: OpusEvent? = null
    var channel: Int = 0

    fun add_digit(value: Int) {
        var current_register = this.register
        if (current_register == null) {
            this.register = OpusEvent(
                value,
                this.RADIX,
                channel,
                false
            )
        } else if (current_register.relative) {
            this.register = OpusEvent(
                current_register.note * value,
                current_register.radix,
                current_register.channel,
                true
            )
        } else {
            this.register = OpusEvent(
                (current_register.note * current_register.radix) + value,
                current_register.radix,
                current_register.channel,
                false
            )
        }
    }

    fun is_ready(): Boolean {
        var register = this.register
        return register != null && (register.relative || register.note >= this.RADIX)
    }

    fun clear() {
        this.register = null
    }

    fun fetch(): OpusEvent? {
        var output = this.register
        this.register = null
        return output
    }

    fun relative_add_entry() {
        this.register = OpusEvent(1, this.RADIX, this.channel, true)
    }

    fun relative_subtract_entry() {
        this.register = OpusEvent(-1, this.RADIX, this.channel, true)
    }

    fun relative_downshift_entry() {
        this.register = OpusEvent(-1 * this.RADIX, this.RADIX, this.channel, true)
    }

    fun relative_upshift_entry() {
        this.register = OpusEvent(this.RADIX, this.RADIX, this.channel, true)
    }

    fun set_channel(channel: Int) {
        this.channel = channel
    }
}
